### ISSUE TITLE
About Loss of InfoNCE and Cluter_results 

### ISSUE THREAD

**Issue by lsyysl9711** [salesforce/PCL#22]
Hi, 

1. I notice that the labels created in InfoNCE loss is always a zero-vector:(https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/pcl/builder.py#L163) 
I think this is wrong since otherwise the loss will always be zero. Did I mis-understand the codes?

2. In creating the Custer_Result dictionary, I found that only eval dataset was involved into consideration:
(https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/main_pcl.py#L299)
So what is the motivation behind this operation, I think we should run it on training set.

**Comment by lerogo** [comment]
> Hi,
> 
> 1. I notice that the labels created in InfoNCE loss is always a zero-vector:(https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/pcl/builder.py#L163
>    )
>    I think this is wrong since otherwise the loss will always be zero. Did I mis-understand the codes?
> 2. In creating the Custer_Result dictionary, I found that only eval dataset was involved into consideration:
>    (https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/main_pcl.py#L299
>    )
>    So what is the motivation behind this operation, I think we should run it on training set.

Same question.

**Comment by Volibear1234** [comment]
> Hi,
> 
> 1. I notice that the labels created in InfoNCE loss is always a zero-vector:(https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/pcl/builder.py#L163
>    )
>    I think this is wrong since otherwise the loss will always be zero. Did I mis-understand the codes?
> 2. In creating the Custer_Result dictionary, I found that only eval dataset was involved into consideration:
>    (https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/main_pcl.py#L299
>    )
>    So what is the motivation behind this operation, I think we should run it on training set.

For the first question, you can refer to the moco_v1 code, where they directly use cross-entropy to implement InfoNCE. But for the second question, I still don't know.

**Comment by Volibear1234** [comment]
> Hi,
> 
> 1. I notice that the labels created in InfoNCE loss is always a zero-vector:(https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/pcl/builder.py#L163
>    )
>    I think this is wrong since otherwise the loss will always be zero. Did I mis-understand the codes?
> 2. In creating the Custer_Result dictionary, I found that only eval dataset was involved into consideration:
>    (https://github.com/salesforce/PCL/blob/964da1fb7c0546e8ce55627fa3c0debde4b7e456/main_pcl.py#L299
>    )
>    So what is the motivation behind this operation, I think we should run it on training set.

For the first question, you can refer to the moco_v1 code, where they use cross-entropy directly for InfoNCE. As for the second question, they use the eval_dataset as negative prototypes, and in the line of code:

```python
output, target, output_proto, target_proto = model(im_q = images[0], im_k = images[1],
                                                   cluster_result = cluster_result, index = index)
```

the passed `index` is from the train_loader, so it still computes based on the train_dataset.
